### ISSUE TITLE
Pick target constructor via BeanMapping

### ISSUE THREAD

**Issue by filiphr** [mapstruct/mapstruct#2080]
A continuation of PR #2074.

Currently the following mechanism is used when there is more than one constructor on the target bean:

* If a parameterless constructor exists then this constructor would be used
* If a constructor is annotated with an annotation named `@Default` (from any package) then this constructor would be used

Based on feedback from users we can offer to define the target constructor parameters in order to pick the appropriate constructor to be used.

e.g.
A `Person` with 2 constructors: 

```
public class Person {


    public Person(String name) {
        // initialize
    }

    public Person(String name, String surname) {
        // initialize
    }
}
```

To pick a constructor something like the following can be used:

```
@BeanMapping(targetConstructorProperties = { "name", "surname" }
```

**Comment by chris922** [comment]
Good idea! Defining the target constructor by parameter names is simple, but MapStruct should consider cases where multiple constructors have the same parameter names, e. g. :

```java
public Foo(String name, String age) { ... }
public Foo(String name, int age) { ... }
```

So maybe better define the target constructor by types or support both and force the latter one in case multiple constructors have the same param-names?

Following two approaches in case MapStruct should support both:

**Approach 1:**

Providing two properties _(as always: namings could be different)_:
1. `targetConstructorProperties`: Define by param names (as `String`)
2. `targetConstructorPropertyTypes`: Define by param types (as `Class`)

If both are defined only `targetConstructorPropertyTypes` MapStruct should fail.

1. If `targetConstructorPropertyTypes` is defined a constructor having exactly the given types in the given order is used
2. Else if `targetConstructorProperties` is defined a constructor having the named properties is used (in any order) as long as there is only one constructor having these properties
3. Else if `targetConstructorProperties` is defined and there are multiple constructors choose the one where the order matches as long as there aren't multiple constructors having the same param names in the correct order
4. Else if `targetConstructorProperties` is defined, multiple constructors are suitable having the same param names in the same order then fail (and refer in the message to `targetConstructorPropertyTypes`)

**Approach 2:**

Only stick to `targetConstructorProperties` where the param names (as `String`) will be defined. In case of multiple constructors having the same named parameters it is possible to "prefix" the param name with a class type, e. g:

```java
@BeanMapping(targetConstructorProperties = { "java.lang.String name", "surname" }
```
To keep it simple we could set `java.lang.` package prefix as optional, so that just `String name` works, but for classes from other packages the package prefix is still required:

```java
@BeanMapping(targetConstructorProperties = { "String name", "com.example.Name surname" }
```

As mentioned in approach 1 first ignore the order, if multiple suitable constructors will be found choose the one where the order matches.

**Approach 3:**

Use `targetConstructorProperties` and define a second annotation `Property` that has the property name as a value and optional a `type` property (as `Class`):

```java
@BeanMapping(targetConstructorProperties = { @Property("name"), @Property(type = com.example.Name.class, value = "surname" }
```

As mentioned in approach 1 first ignore the order, if multiple suitable constructors will be found choose the one where the order matches.
